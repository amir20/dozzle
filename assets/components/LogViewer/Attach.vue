<template>
  <aside>
    <header class="flex items-center gap-4">
      <h1 class="text-2xl max-md:hidden">{{ container.name }}</h1>
      <h2 class="text-sm"><DistanceTime :date="container.created" /></h2>
    </header>

    <div class="mt-8 flex flex-col gap-2">
      <section>
        <div ref="terminal"></div>
      </section>
    </div>
  </aside>
</template>

<script setup lang="ts">
import { Container } from "@/models/Container";
import "@xterm/xterm/css/xterm.css";
const { container } = defineProps<{ container: Container }>();

const { Terminal } = await import("@xterm/xterm");

const terminal = useTemplateRef<HTMLDivElement>("terminal");
const term = new Terminal({
  cursorBlink: true,
  cursorStyle: "block",
});
const ws = new WebSocket(withBase(`/api/hosts/${container.host}/containers/${container.id}/attach`));

onMounted(() => {
  term.open(terminal.value!);
});

onUnmounted(() => {
  term.dispose();
  ws.close();
  console.log("WebSocket closed");
});

ws.onopen = () => {
  term.writeln("Attached to container ðŸš€");
  term.onData((data) => {
    ws.send(data);
  });
};

ws.onmessage = (event) => term.write(event.data);
</script>
<style scoped></style>
