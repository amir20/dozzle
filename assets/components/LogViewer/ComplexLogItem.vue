<template>
  <div class="group/item relative flex w-full gap-x-2" @click="expandToggle()">
    <label class="swap swap-rotate invisible absolute -left-4 top-0.5 size-4 group-hover/item:visible">
      <input type="checkbox" v-model="expanded" @click="expandToggle()" />
      <material-symbols:expand-all-rounded class="swap-off text-secondary" />
      <material-symbols:collapse-all-rounded class="swap-on text-secondary" />
    </label>
    <div v-if="showContainerName">
      <ContainerName :id="logEntry.containerID" />
    </div>
    <div v-if="showStd">
      <LogStd :std="logEntry.std" />
    </div>
    <div v-if="showTimestamp">
      <LogDate :date="logEntry.date" />
    </div>
    <div class="flex">
      <LogLevel :level="logEntry.level" />
    </div>
    <div>
      <ul class="fields cursor-pointer space-x-4" :class="{ expanded }">
        <li v-for="(value, name) in validValues" :key="name">
          <span class="text-light">{{ name }}=</span><span class="font-bold" v-if="value === null">&lt;null&gt;</span>
          <template v-else-if="Array.isArray(value)">
            <span class="font-bold" v-html="markSearch(JSON.stringify(value))"> </span>
          </template>
          <span class="font-bold" v-html="markSearch(value)" v-else></span>
        </li>
        <li class="text-light" v-if="Object.keys(validValues).length === 0">all values are hidden</li>
      </ul>
      <FieldList :fields="logEntry.unfilteredMessage" :expanded="expanded" :visible-keys="visibleKeys" />
    </div>
    <LogMessageActions
      class="duration-250 absolute -right-1 opacity-0 transition-opacity delay-150 group-hover/entry:opacity-100"
      :message="() => JSON.stringify(logEntry.message)"
      :log-entry="logEntry"
    />
  </div>
</template>
<script lang="ts" setup>
import { type ComplexLogEntry } from "@/models/LogEntry";

const { markSearch } = useSearchFilter();

const { logEntry, showContainerName = false } = defineProps<{
  logEntry: ComplexLogEntry;
  visibleKeys: string[][];
  showContainerName?: boolean;
}>();

const [expanded, expandToggle] = useToggle();

const validValues = computed(() => {
  return Object.fromEntries(Object.entries(logEntry.message).filter(([_, value]) => value !== undefined));
});
</script>

<style lang="postcss" scoped>
.text-light {
  @apply text-base-content/70;
}

.fields {
  li {
    @apply inline-flex;
  }
}
</style>
