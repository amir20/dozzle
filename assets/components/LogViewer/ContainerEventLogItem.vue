<template>
  <div class="flex w-full flex-col">
    <div class="relative flex items-start gap-x-2">
      <ContainerName class="flex-none" :id="logEntry.containerID" v-if="showContainerName" />
      <LogDate :date="logEntry.date" v-if="showTimestamp" />
      <LogLevel class="flex" />
      <div class="whitespace-pre-wrap" :data-event="logEntry.event" v-html="logEntry.message"></div>
    </div>
    <div class="alert alert-info mt-8 w-auto flex-none font-sans text-[1rem] md:mx-auto md:w-1/2" v-if="followEligible">
      <carbon:information class="size-6 shrink-0 stroke-current" />
      <div>
        <h3 class="text-lg font-bold">{{ $t("alert.similar-container-found.title") }}</h3>
        {{ $t("alert.similar-container-found.message", { containerId: nextContainer.id }) }}
      </div>
      <div>
        <TimedButton v-if="automaticRedirect" class="btn-primary btn-sm" @finished="redirectNow()">{{
          $t("button.cancel")
        }}</TimedButton>
        <router-link
          :to="{ name: '/container/[id]', params: { id: nextContainer.id } }"
          class="btn btn-primary btn-sm"
          v-else
        >
          {{ $t("button.redirect") }}
        </router-link>
      </div>
    </div>
  </div>
</template>
<script lang="ts" setup>
import { ContainerEventLogEntry } from "@/models/LogEntry";
const router = useRouter();
const { showToast } = useToast();
const { t } = useI18n();

const { logEntry } = defineProps<{
  logEntry: ContainerEventLogEntry;
  showContainerName?: boolean;
}>();

const { containers } = useLoggingContext();

const store = useContainerStore();

const { containers: allContainers } = storeToRefs(store);

const nextContainer = computed(
  () =>
    [
      ...allContainers.value.filter(
        (c) =>
          c.host === containers.value[0].host &&
          c.created > logEntry.date &&
          c.name === containers.value[0].name &&
          c.state === "running",
      ),
    ].sort((a, b) => +a.created - +b.created)[0],
);

const followEligible = computed(
  () =>
    router.currentRoute.value.name === "/container/[id]" &&
    logEntry.event === "container-stopped" &&
    containers.value.length === 1 &&
    nextContainer.value !== undefined,
);

function redirectNow() {
  showToast(
    {
      title: t("alert.redirected.title"),
      message: t("alert.redirected.message", { containerId: nextContainer.value?.id }),
      type: "info",
    },
    { expire: 5000 },
  );
  router.push({ name: "/container/[id]", params: { id: nextContainer.value?.id } });
}
</script>

<style lang="postcss" scoped>
[data-event="container-stopped"] {
  @apply text-red;
}
[data-event="container-started"] {
  @apply text-green;
}
</style>
