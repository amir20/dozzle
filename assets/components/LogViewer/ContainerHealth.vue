<template>
  <div class="icon is-small health" :health="health" v-if="health" :title="health">
    <cil:check-circle v-if="health == 'healthy'" />
    <cil:x-circle v-else-if="health == 'unhealthy'" />
    <cil:circle v-else />
  </div>
</template>

<script lang="ts" setup>
import { ContainerHealth } from "@/types/Container";

defineProps<{
  health: ContainerHealth | undefined;
}>();
</script>

<style lang="scss" scoped>
.health {
  &[health="unhealthy"] {
    color: var(--red-color);
  }

  &[health="healthy"] {
    color: var(--green-color);
  }
}
</style>
