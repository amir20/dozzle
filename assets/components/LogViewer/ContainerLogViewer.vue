<template>
  <log-viewer :messages="filtered" :last-selected-item="lastSelectedItem" :visible-keys="visibleKeys" />
</template>

<script lang="ts" setup>
import { useRouteHash } from "@vueuse/router";

import { type JSONObject, LogEntry } from "@/models/LogEntry";

const props = defineProps<{
  messages: LogEntry<string | JSONObject>[];
}>();

const { container } = useContainerContext();

const visibleKeys = persistentVisibleKeys(container);

const { filteredPayload } = useVisibleFilter(visibleKeys);
const { filteredMessages } = useSearchFilter();

const { messages } = toRefs(props);
const visible = filteredPayload(messages);
const filtered = filteredMessages(visible);

const { lastSelectedItem } = useLogSearchContext() as {
  lastSelectedItem: Ref<LogEntry<string | JSONObject> | undefined>;
};
const routeHash = useRouteHash();
watch(
  routeHash,
  (hash) => {
    if (hash) {
      document.querySelector(`[data-key="${hash.substring(1)}"]`)?.scrollIntoView({ block: "center" });
    }
  },
  { immediate: true, flush: "post" },
);
</script>
<style scoped lang="postcss"></style>
