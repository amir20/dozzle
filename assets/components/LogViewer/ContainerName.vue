<template>
  <div class="relative block w-40 overflow-hidden rounded px-1.5 text-center text-sm text-white">
    <div class="random-color absolute inset-0 brightness-75"></div>
    <div class="direction-rtl relative truncate">{{ containerNames[id] }}</div>
  </div>
</template>
<script lang="ts" setup>
const containerStore = useContainerStore();
const { containerNames } = storeToRefs(containerStore);

const { id } = defineProps<{
  id: string;
}>();

const colors = [
  "#FF0000",
  "#FF7F00",
  "#FFFF00",
  "#00FF00",
  "#00FFFF",
  "#0000FF",
  "#FF00FF",
  "#FF007F",
  "#32CD32",
  "#40E0D0",
  "#800080",
  "#FFD700",
  "#FF4040",
  "#4B0082",
  "#008080",
  "#E6E6FA",
];

const color = computed(() => colors[Math.abs(hashCode(id)) % colors.length]);
</script>

<style lang="postcss" scoped>
.random-color {
  background-color: v-bind(color);
}

.direction-rtl {
  direction: rtl;
}
</style>
