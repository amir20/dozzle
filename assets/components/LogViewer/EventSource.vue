<template>
  <InfiniteLoader :onLoadMore="fetchMore" :enabled="messages.length > 50"></InfiniteLoader>
  <slot :messages="messages"></slot>
</template>

<script lang="ts" setup generic="T">
import { LogStreamSource } from "@/composable/eventStreams";

const loadingMore = defineEmit<[value: boolean]>();

const props = defineProps<{
  streamSource: (t: Ref<T>) => LogStreamSource;
  entity: T;
}>();

const { entity, streamSource } = toRefs(props);

const { messages, loadOlderLogs } = streamSource.value(entity);

const beforeLoading = () => loadingMore(true);
const afterLoading = () => loadingMore(false);

defineExpose({
  clear: () => (messages.value = []),
});

const fetchMore = () => loadOlderLogs({ beforeLoading, afterLoading });
</script>
