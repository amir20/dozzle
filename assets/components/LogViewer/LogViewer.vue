<template>
  <ul class="events" ref="events" :class="{ 'disable-wrap': !softWrap, [size]: true }">
    <li
      v-for="(item, index) in filtered"
      :key="item.id"
      :data-key="item.id"
      :class="{ selected: toRaw(item) === toRaw(lastSelectedItem) }"
    >
      <div class="line-options" v-show="isSearching()">
        <dropdown-menu :class="{ 'is-last': index === filtered.length - 1 }" class="is-top minimal">
          <a class="dropdown-item" @click="handleJumpLineSelected($event, item)" :href="`#${item.id}`">
            <div class="level is-justify-content-start">
              <div class="level-left">
                <div class="level-item">
                  <cil-find-in-page class="mr-4" />
                </div>
              </div>
              <div class="level-right">
                <div class="level-item">Jump to Context</div>
              </div>
            </div>
          </a>
        </dropdown-menu>
      </div>
      <div class="line">
        <span class="date" v-if="showTimestamp"> <relative-time :date="item.date"></relative-time></span>
        <component :is="item.getComponent()" :log-entry="item" :visible-keys="visibleKeys.value"></component>
      </div>
    </li>
  </ul>
</template>

<script lang="ts" setup>
import { type ComputedRef, toRaw } from "vue";
import { useRouteHash } from "@vueuse/router";
import { type Container } from "@/types/Container";
import { type JSONObject, type LogEntry } from "@/models/LogEntry";

const props = defineProps<{
  messages: LogEntry<string | JSONObject>[];
}>();

const { messages } = toRefs(props);
let visibleKeys = persistentVisibleKeys(inject("container") as ComputedRef<Container>);

const { filteredPayload } = useVisibleFilter(visibleKeys);
const { filteredMessages, resetSearch, isSearching } = useSearchFilter();

const visible = filteredPayload(messages);
const filtered = filteredMessages(visible);

const events = ref<HTMLElement>();
let lastSelectedItem = ref<LogEntry<string | JSONObject>>();

function handleJumpLineSelected(e: Event, item: LogEntry<string | JSONObject>) {
  lastSelectedItem.value = item;
  resetSearch();
}

const routeHash = useRouteHash();
watch(
  routeHash,
  (hash) => {
    document.querySelector(`[data-key="${hash.substring(1)}"]`)?.scrollIntoView({ block: "center" });
  },
  { immediate: true, flush: "post" }
);
</script>
<style scoped lang="scss">
.events {
  padding: 1em 0;
  font-family: SFMono-Regular, Consolas, Liberation Mono, monaco, Menlo, monospace;

  &.disable-wrap {
    .line {
      white-space: nowrap;
    }
  }

  & > li {
    display: flex;
    word-wrap: break-word;
    padding: 0.2em 1em;
    &:last-child {
      scroll-snap-align: end;
      scroll-margin-block-end: 5rem;
    }
    &:nth-child(odd) {
      background-color: rgba(125, 125, 125, 0.08);
    }

    &.selected .date {
      background-color: var(--menu-item-active-background-color);

      color: var(--text-color);
    }
    &.selected > .date {
      background-color: white;
    }
    & > .line {
      margin: auto 0;
      width: 100%;
      display: flex;
    }
    & > .line-options {
      display: flex;
      flex-direction: row-reverse;
      margin-right: 1em;
    }
  }

  &.small {
    font-size: 60%;
  }

  &.medium {
    font-size: 80%;
  }

  &.large {
    font-size: 120%;
  }
}

@media (prefers-color-scheme: dark) {
  .date {
    background-color: #262626;
    color: #258ccd;
  }
}

[data-theme="dark"] {
  .date {
    background-color: #262626;
    color: #258ccd;
  }
}

@media (prefers-color-scheme: light) {
  .date {
    background-color: #f0f0f0;
    color: #009900;
  }
}

[data-theme="light"] {
  .date {
    background-color: #f0f0f0;
    color: #009900;
  }
}

.date {
  padding-left: 5px;
  padding-right: 5px;
  border-radius: 3px;
}

:deep(mark) {
  border-radius: 2px;
  background-color: var(--secondary-color);
  animation: pops 200ms ease-out;
  display: inline-block;
}

@keyframes pops {
  0% {
    transform: scale(1.5);
  }
  100% {
    transform: scale(1.05);
  }
}
</style>
