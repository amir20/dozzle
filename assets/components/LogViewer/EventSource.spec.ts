import { createTestingPinia } from "@pinia/testing";
import { mount } from "@vue/test-utils";
import { useSearchFilter } from "@/composable/search";
import { settings } from "@/stores/settings";
// @ts-ignore
import EventSource, { sources } from "eventsourcemock";
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { computed, nextTick } from "vue";
import { createI18n } from "vue-i18n";
import { createRouter, createWebHistory } from "vue-router";
import { default as Component } from "./EventSource.vue";
import LogViewer from "@/components/LogViewer/LogViewer.vue";
import { Container } from "@/models/Container";

vi.mock("@/stores/config", () => ({
  __esModule: true,
  default: { base: "", hosts: [{ name: "localhost", id: "localhost" }] },
  withBase: (path: string) => path,
}));

/**
 * @vitest-environment jsdom
 */
describe("<ContainerEventSource />", () => {
  const search = useSearchFilter();

  beforeEach(() => {
    global.EventSource = EventSource;
    // @ts-ignore
    window.scrollTo = vi.fn();
    global.IntersectionObserver = vi.fn().mockImplementation(() => ({
      observe: vi.fn(),
      disconnect: vi.fn(),
    }));
    vi.useFakeTimers();
    vi.setSystemTime(1560336942459);
  });

  afterEach(() => {
    vi.restoreAllMocks();
    vi.useRealTimers();
  });

  function createLogEventSource(
    {
      searchFilter = "",
      hourStyle = "auto",
    }: { searchFilter?: string | undefined; hourStyle?: "auto" | "24" | "12" } = {
      hourStyle: "auto",
    },
  ) {
    settings.value.hourStyle = hourStyle;
    search.searchFilter.value = searchFilter;
    if (searchFilter) {
      search.showSearch.value = true;
    }

    const router = createRouter({
      history: createWebHistory("/"),
      routes: [
        {
          path: "/",
          component: {
            template: "Test from createLogEventSource",
          },
        },
      ],
    });

    return mount(Component<Container>, {
      global: {
        plugins: [router, createTestingPinia({ createSpy: vi.fn }), createI18n({})],
        components: {
          LogViewer,
        },
        provide: {
          scrollingPaused: computed(() => false),
          [loggingContextKey as symbol]: {
            containers: computed(() => [{ id: "abc", image: "test:v123", host: "localhost" }]),
            streamConfig: reactive({ stdout: true, stderr: true }),
          },
        },
      },
      slots: {
        default: `
        <template #scoped="params"><LogViewer :messages="params.messages" :show-container-name="false" :visible-keys="[]" /></template>
        `,
      },
      props: {
        streamSource: useContainerStream,
        entity: new Container("abc", new Date(), "image", "name", "command", "localhost", {}, "status", "created", []),
      },
    });
  }

  const sourceUrl = "/api/hosts/localhost/containers/abc/logs/stream?stdout=1&stderr=1";

  test("renders correctly", async () => {
    const wrapper = createLogEventSource();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should connect to EventSource", async () => {
    const wrapper = createLogEventSource();
    sources[sourceUrl].emitOpen();
    expect(sources[sourceUrl].readyState).toBe(1);
    wrapper.unmount();
  });

  test("should close EventSource", async () => {
    const wrapper = createLogEventSource();
    sources[sourceUrl].emitOpen();
    wrapper.unmount();
    expect(sources[sourceUrl].readyState).toBe(2);
  });

  test("should parse messages", async () => {
    const wrapper = createLogEventSource();
    sources[sourceUrl].emitOpen();
    sources[sourceUrl].emitMessage({
      data: `{"ts":1560336942459, "m":"This is a message.", "id":1}`,
    });

    vi.runAllTimers();
    await nextTick();

    // @ts-ignore
    const [message, _] = wrapper.vm.messages;
    expect(message).toMatchSnapshot();
  });

  describe("render html correctly", () => {
    test("should render messages", async () => {
      const wrapper = createLogEventSource();
      sources[sourceUrl].emitOpen();
      sources[sourceUrl].emitMessage({
        data: `{"ts":1560336942459, "m":"This is a message.", "id":1}`,
      });

      vi.runAllTimers();
      await nextTick();

      expect(wrapper.find("ul.events").html()).toMatchSnapshot();
    });

    test("should render messages with html entities", async () => {
      const wrapper = createLogEventSource();
      sources[sourceUrl].emitOpen();
      sources[sourceUrl].emitMessage({
        data: `{"ts":1560336942459, "m":"<test>foo bar</test>", "id":1}`,
      });

      vi.runAllTimers();
      await nextTick();

      expect(wrapper.find("ul.events").html()).toMatchSnapshot();
    });

    test("should render dates with 12 hour style", async () => {
      const wrapper = createLogEventSource({ hourStyle: "12" });
      sources[sourceUrl].emitOpen();
      sources[sourceUrl].emitMessage({
        data: `{"ts":1560336942459, "m":"<test>foo bar</test>", "id":1}`,
      });

      vi.runAllTimers();
      await nextTick();

      expect(wrapper.find("ul.events").html()).toMatchSnapshot();
    });

    test("should render dates with 24 hour style", async () => {
      const wrapper = createLogEventSource({ hourStyle: "24" });
      sources[sourceUrl].emitOpen();
      sources[sourceUrl].emitMessage({
        data: `{"ts":1560336942459, "m":"<test>foo bar</test>", "id":1}`,
      });

      vi.runAllTimers();
      await nextTick();

      expect(wrapper.find("ul.events").html()).toMatchSnapshot();
    });

    test("should render messages with filter", async () => {
      const wrapper = createLogEventSource({ searchFilter: "test" });
      sources[sourceUrl].emitOpen();
      sources[sourceUrl].emitMessage({
        data: `{"ts":1560336942459, "m":"foo bar", "id":1}`,
      });
      sources[sourceUrl].emitMessage({
        data: `{"ts":1560336942459, "m":"test bar", "id":2}`,
      });

      vi.runAllTimers();
      await nextTick();

      expect(wrapper.find("ul.events").html()).toMatchSnapshot();
    });
  });
});
