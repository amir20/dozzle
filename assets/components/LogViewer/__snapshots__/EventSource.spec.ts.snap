// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ContainerEventSource /> > render html correctly > should render dates with 12 hour style 1`] = `
"<ul data-v-cf9ff940="" class="events group pt-4 medium" show-container-name="false">
  <li data-v-cf9ff940="" data-key="1" data-time="1560336942459" class="group/entry">
    <div data-v-a49e52d4="" data-v-cf9ff940="" class="relative flex w-full items-start gap-x-2 group-[.compact]:items-stretch">
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <div class="tag bg-base-100 inline-flex items-center justify-center rounded-sm px-2 py-[0.2em] [[size='small']]:text-[0.8rem] items-start! shrink-0 select-none" size="small">
        <div class="inline-flex gap-2 text-blue whitespace-nowrap"><time datetime="2019-06-12T10:55:42.459Z" class="max-md:hidden">06/12/2019</time><time datetime="2019-06-12T10:55:42.459Z">10:55:42 AM</time></div>
      </div>
      <div data-v-e625cddd="" class="mt-1.5 size-2.5 flex-none rounded-lg flex select-none"></div>
      <div data-v-a49e52d4="" class="log-wrapper [word-break:break-word] whitespace-pre-wrap group-[.disable-wrap]:whitespace-nowrap">foo bar</div>
      <div data-v-a49e52d4="" class="flex gap-2 absolute -right-1 opacity-0 transition-opacity delay-150 duration-250 group-hover/entry:opacity-100">
        <!--v-if-->
      </div>
    </div>
  </li>
</ul>"
`;

exports[`<ContainerEventSource /> > render html correctly > should render dates with 24 hour style 1`] = `
"<ul data-v-cf9ff940="" class="events group pt-4 medium" show-container-name="false">
  <li data-v-cf9ff940="" data-key="1" data-time="1560336942459" class="group/entry">
    <div data-v-a49e52d4="" data-v-cf9ff940="" class="relative flex w-full items-start gap-x-2 group-[.compact]:items-stretch">
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <div class="tag bg-base-100 inline-flex items-center justify-center rounded-sm px-2 py-[0.2em] [[size='small']]:text-[0.8rem] items-start! shrink-0 select-none" size="small">
        <div class="inline-flex gap-2 text-blue whitespace-nowrap"><time datetime="2019-06-12T10:55:42.459Z" class="max-md:hidden">06/12/2019</time><time datetime="2019-06-12T10:55:42.459Z">10:55:42</time></div>
      </div>
      <div data-v-e625cddd="" class="mt-1.5 size-2.5 flex-none rounded-lg flex select-none"></div>
      <div data-v-a49e52d4="" class="log-wrapper [word-break:break-word] whitespace-pre-wrap group-[.disable-wrap]:whitespace-nowrap">foo bar</div>
      <div data-v-a49e52d4="" class="flex gap-2 absolute -right-1 opacity-0 transition-opacity delay-150 duration-250 group-hover/entry:opacity-100">
        <!--v-if-->
      </div>
    </div>
  </li>
</ul>"
`;

exports[`<ContainerEventSource /> > render html correctly > should render messages 1`] = `
"<ul data-v-cf9ff940="" class="events group pt-4 medium" show-container-name="false">
  <li data-v-cf9ff940="" data-key="1" data-time="1560336942459" class="group/entry">
    <div data-v-a49e52d4="" data-v-cf9ff940="" class="relative flex w-full items-start gap-x-2 group-[.compact]:items-stretch">
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <div class="tag bg-base-100 inline-flex items-center justify-center rounded-sm px-2 py-[0.2em] [[size='small']]:text-[0.8rem] items-start! shrink-0 select-none" size="small">
        <div class="inline-flex gap-2 text-blue whitespace-nowrap"><time datetime="2019-06-12T10:55:42.459Z" class="max-md:hidden">06/12/2019</time><time datetime="2019-06-12T10:55:42.459Z">10:55:42 AM</time></div>
      </div>
      <div data-v-e625cddd="" class="mt-1.5 size-2.5 flex-none rounded-lg flex select-none"></div>
      <div data-v-a49e52d4="" class="log-wrapper [word-break:break-word] whitespace-pre-wrap group-[.disable-wrap]:whitespace-nowrap">This is a message.</div>
      <div data-v-a49e52d4="" class="flex gap-2 absolute -right-1 opacity-0 transition-opacity delay-150 duration-250 group-hover/entry:opacity-100">
        <!--v-if-->
      </div>
    </div>
  </li>
</ul>"
`;

exports[`<ContainerEventSource /> > should parse messages 1`] = `
SimpleLogEntry {
  "_message": "This is a message.",
  "containerID": undefined,
  "date": 2019-06-12T10:55:42.459Z,
  "id": 1,
  "level": undefined,
  "position": undefined,
  "std": "stderr",
}
`;
